### DISCUSSION TITLE
Can I use match() on upsert()? [Getting advice]

### DISCUSSION THREAD

**Discussion by aditodkar** [supabase/supabase#4546] [Questions] [answered]
I have table which has book_id, author_id, token and id(auto generated uuidv4) columns. Now I want to update token data based on author_id and book_id match and if that book_id and author_id is not present then I want to insert data into table. I have done it in below code but is there any better way to do it?

```
const { data, error } = await supabase.from('table').select().match({ book_id: bookId, author_id: authorId });
        if (error) {
            console.log("error", error)
            return res.status(424).send({ message: 'failed' })
        }

        if (data.length > 0) {
            const { error } = await supabase.from('table').update({ token: token }).match({ book_id: bookId, author_id: authorId });
            if (error) {
                console.log("error", error)
                return res.status(424).send({ message: 'failed' })
            }
        } else {
            const { error } = await supabase.from('table').insert({ token: token, book_id: bookId, author_id: authorId });
            if (error) {
                console.log("error", error)
                return res.status(424).send({ message: 'failed' })
            }
        }
```

**Comment by GaryAustin1** [comment, marked as answer]
Edit: I had suggested add a concatenated unique column here, but Postgres actual has a Unique constraint that works on multiple columns see futher on...

If you created a column (maybe generated if you don't want to pass extra data) combining author_id and book_id and put a unique constraint on it then upsert using onConflict would probably do what you want.  Otherwise I'm sure you could write an rpc function to avoid 2  database trips.

**Reply by aditodkar** [reply]
If possible can you share any example? Did not find anything in docs.

**Reply by GaryAustin1** [reply]
So in looking for Postgres example, I've found this which looks even better:
```sql
ALTER TABLE table ADD UNIQUE (book_id, author_id)
```
What that does is enforce that those two columns can only be inserted once if they both match without actually adding an extra column with them both concatenated.
Then if you follow the upsert example for Supabase the onConflict SHOULD work to update rather than insert if they both exist in a pair.
I've not tried this though....

**Reply by aditodkar** [reply]
Thanks a lot got it,

**Reply by dom-the-dev** [reply]
@aditodkar could you share your new solution? I'm sitting on a similiar issue.

**Reply by vincentjflorio** [reply]
@aditodkar 

I had a slightly different SQL editor line:
`ALTER TABLE favorites remove UNIQUE (user_id, asset_id)`

and then
`const { data, error } = await supabaseClient.from('favorites')
            .upsert(preparedUpsert, {onConflict : 'user_id,asset_id'})`
            
it may be worth noting that the SQL editor burped back a gangly key that looks like **user_id_asset_id_pfkey** but I stupidly didn't copy it down. It seems to work, though, writing the comma version in the API call. Definitely changes the table correct across two different users, and no console errors.

**Reply by dalmo3** [reply]
@vincentjflorio thanks for sharing this. It's undocumented that you can pass comma-separated column names to `onConflict`.

**Reply by steve-chavez** [reply]
@dalmo3 @vincentjflorio A PR for the onConflict with multiple columns would be great. 

https://supabase.com/docs/reference/javascript/upsert

**Reply by ericrav** [reply]
> @aditodkar
> 
> I had a slightly different SQL editor line: `ALTER TABLE favorites remove UNIQUE (user_id, asset_id)`

Should this line be:
```sql
ALTER TABLE favorites add UNIQUE (user_id, asset_id)
```

Passing comma-separated values to `onConflict` only worked for me if I already had a unique constraint on those fields.

(very helpful though! thanks)
